### PR TITLE
add OFFSET, LIMIT and ORDER BY keywords

### PR DESCRIPTION
**Pull request by hmmr**
RTS-733, RTS-870, RTS-1006

Required by https://github.com/basho/riak_kv/pull/1479, which see.

### CODE REVIEW

File: src/riak_ql_parser.yrl [hunk 1]
```diff
@@ -566,8 +604,36 @@ make_insert({identifier, Table}, Fields, Values) ->
...
+%%  non-null values. If neither is specified, the default behavior is
+%%  NULLS LAST when ASC is specified or implied, and NULLS FIRST when
+%%  DESC is specified (thus, the default is to act as though nulls are
+%%  larger than non-nulls)."
+make_ordby_item(Fld, {asc, _} = Dir, undefined) ->
+    {Fld, Dir, {nulls_last, <<"nulls last">>}};
+make_ordby_item(Fld, {desc, _} = Dir, undefined) ->
+    {Fld, Dir, {nulls_first, <<"nulls first">>}};
+make_ordby_item(Fld, Dir, Nulls) ->
+    {Fld,
```
Comment (gordonguthrie) [inline src/riak_ql_parser.yrl]: I find this style of evaluating statements to substitute into tuples to be very ugly

The reason is that I have to evaluate the code in my head to work out what you are going to return

If you evaluated each if into a named variable then I could read the function head, read the return variable and know what the function was doing and then drop in to each variable and inspect it at my leisure...

File: test/parser_select_tests.erl [hunk 2]
```diff
@@ -163,9 +163,60 @@ group_by_two_fields_test() ->
         {select, [
                   {tables, <<"mytab">>},
                   {fields, [{identifier, [<<"a">>]}, {identifier, [<<"b">>]}]},
-                  {limit,  []},
                   {where,  [{'=', <<"a">>, {integer, 1}}]},
-                  {group_by, [{identifier, <<"a">>}, {identifier, <<"b">>}]}
+                  {group_by, [{identifier, <<"a">>}, {identifier, <<"b">>}]},
+                  {limit,undefined},{offset,undefined},{order_by,undefined}
```
Comment (gordonguthrie) [inline test/parser_select_tests.erl]: Go vertical or go horizontal but don't mix and match...

File: test/parser_select_tests.erl [hunk 3]
```diff
@@ -163,9 +163,60 @@ group_by_two_fields_test() ->
...
+        {select, [
+                  {tables, <<"mytab">>},
+                  {fields, [{identifier, [<<"a">>]}, {identifier, [<<"b">>]}]},
+                  {where,  [{'=', <<"a">>, {integer, 1}}]},
+                  {group_by, []},
+                  {limit,undefined},{offset,undefined},
+                  {order_by,[{<<"a">>, desc, nulls_first}, {<<"b">>, asc, nulls_last}]}
+                 ]},
+        riak_ql_parser:ql_parse(riak_ql_lexer:get_tokens(Query_sql))
+    ).
```
Comment (gordonguthrie) [inline test/parser_select_tests.erl:211]: No offset tests, no limit tests, no tests with GROUP as well as limit...

File: src/riak_ql_parser.yrl [hunk 4]
```diff
@@ -135,17 +148,36 @@ Statement -> StatementWithoutSemicolon semicolon : '$1'.
 
 GroupBy -> group by Fields: {group_by, '$3'}.
 
+SortDirection -> asc: '$1'.
```
Comment (gordonguthrie) [inline src/riak_ql_parser.yrl]: Tokens are not in concordance with SQL Spec

screen shot 2016-10-18 at 15 26 50

File: src/riak_ql_ddl.erl [hunk 5]
```diff
@@ -274,19 +277,20 @@ syntax_error_to_msg2({operator_type_mismatch, Fn, Type1, Type2}) ->
...
+               {T2, _Select, _Where, _OrderBy}) when T1 /= T2 ->
     {false, [{bucket_type_mismatch, {T1, T2}}]};
-is_query_valid(Mod, _, {_Table, Selection, Where}) ->
-    ValidSelection = are_selections_valid(Mod, Selection, ?CANTBEBLANK),
+is_query_valid(Mod, _, {_Table, Selection, Where, OrderBy}) ->
+    ValidSelection = check_selections_valid(Mod, Selection, ?CANTBEBLANK),
+    ValidOrderBy   = check_orderby_valid(Mod, OrderBy),
     ValidFilters   = check_filters_valid(Mod, Where),
-    is_query_valid_result(ValidSelection, ValidFilters).
+    is_query_valid_result([ValidSelection, ValidFilters, ValidOrderBy]).
```
Comment (hazen) [inline src/riak_ql_ddl.erl:316]: Doesn't `is_query_valid_result/1` now only expect a list of 2 items?

File: src/riak_ql_ddl.erl [hunk 6]
```diff
@@ -412,11 +424,11 @@ is_compatible_operator('!=', boolean, boolean)-> true;
 is_compatible_operator(_,    boolean, boolean)-> false;
 is_compatible_operator(_,_,_)                 -> true.
 
--spec are_selections_valid(module(), [selection()], boolean()) ->
+-spec check_selections_valid(module(), [selection()], boolean()) ->
```
Comment (hazen) [inline src/riak_ql_ddl.erl]: The original name matches some of the other verification function names starting with `is_` for example

File: src/riak_ql_to_string.erl [hunk 7]
```diff
@@ -92,11 +92,11 @@ op_to_string(Op) ->
 flat_format(Format, Args) ->
     lists:flatten(io_lib:format(Format, Args)).
 
--spec ddl_rec_to_sql(#ddl_v1{}) -> string().
-ddl_rec_to_sql(#ddl_v1{table         = Tb,
-                       fields        = Fs,
-                       partition_key = PK,
-                       local_key     = LK}) ->
+-spec ddl_rec_to_sql(?DDL{}) -> string().
```
Comment (hazen) [inline src/riak_ql_to_string.erl:95]: Good catch. More consistent.

File: src/riak_ql_lexer.xrl [hunk 8]
```diff
@@ -75,22 +82,29 @@ SEMICOLON = (\;)
...
 {LIMIT} : {token, {limit, list_to_binary(TokenChars)}}.
 {NOT} : {token, {not_, list_to_binary(TokenChars)}}.
 {NULL} : {token, {null, list_to_binary(TokenChars)}}.
+{NULLS} : {token, {nulls, list_to_binary(TokenChars)}}.
+{OFFSET} : {token, {offset, list_to_binary(TokenChars)}}.
 {OR} : {token, {or_, list_to_binary(TokenChars)}}.
+{ORDER} : {token, {order, list_to_binary(TokenChars)}}.
 {PRIMARY} : {token, {primary, list_to_binary(TokenChars)}}.
 {QUANTUM} : {token, {quantum, list_to_binary(TokenChars)}}.
 {SELECT} : {token, {select, list_to_binary(TokenChars)}}.
```
Comment (hazen) [inline src/riak_ql_lexer.xrl:111]: We see these warnings:

```
Compiled src/riak_ql_lexer.xrl
src/riak_ql_parser.yrl:16: Warning: non-terminal symbol Buckets not used
src/riak_ql_parser.yrl:113: Warning: terminal symbol is_not_null not used
src/riak_ql_parser.yrl:114: Warning: terminal symbol is_null not used
src/riak_ql_parser.yrl:16: Warning: non-terminal symbol Buckets not used
src/riak_ql_parser.yrl:113: Warning: terminal symbol is_not_null not used
src/riak_ql_parser.yrl:114: Warning: terminal symbol is_null not used
```

Comment (hmmr) [reply src/riak_ql_lexer.xrl:111]: Removed unused tokens.

File: src/riak_ql_ddl.erl (hunk 5 above)
Comment (hmmr) [reply src/riak_ql_ddl.erl:316]: It has been rewritten to accept a list.

File: src/riak_ql_ddl.erl (hunk 6 above)
Comment (hmmr) [reply src/riak_ql_ddl.erl]: Renamed as you suggest.

**Review by gordonguthrie — COMMENTED** [review]
Only general comment was caught by @javajolt - the unused tokens warning

File: src/riak_ql_parser.yrl (hunk 4 above)
Comment (hmmr) [reply src/riak_ql_parser.yrl]: Rewrote the rules using the official idioms. I have to note that there is no rule for the LIMIT [OFFSET] clause, oddly enough, which I had to invent ad hoc.

File: src/riak_ql_parser.yrl (hunk 1 above)
Comment (hmmr) [reply src/riak_ql_parser.yrl]: Now there are `make_ord_spec` and `make_null_spec` functions, to lessen eye strain.

File: test/parser_select_tests.erl (hunk 2 above)
Comment (hmmr) [reply test/parser_select_tests.erl]: Realigned.

File: test/parser_select_tests.erl (hunk 3 above)
Comment (hmmr) [reply test/parser_select_tests.erl:211]: Added some.

**Comment by gordonguthrie** [comment]
+1 beb6981

**Comment by gordonguthrie** [comment]
+1 1c275ea

**Comment by andytill** [comment]
+1 1c275ea

**Pull request merged**